addons/l10n_se/models/res_partner.py

```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import re
+from odoo import api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    orgnumber = fields.Char('Organisational Number')
+
+    @api.onchange('orgnumber')
+    def onchnage_orgnumber(self):
+        def _check_orgnumber(orgnumber):
+            digits = [int(d) for d in re.sub(r'\D', '', orgnumber)][-10:]
+            if len(digits) != 10:
+                return False
+            even_digitsum = sum(x if x < 5 else x - 9 for x in digits[::2])
+            return 0 == sum(digits, even_digitsum) % 10
+        if self.orgnumber:
+            if _check_orgnumber(self.orgnumber):
+                self.vat = 'SE' + self.orgnumber + '01'
```

Comment:
Please, take a look at page 8, explaination 11 in below official pdf from Skatteverket (Swedish Tax Agency), published April 2016 called SKV560B, 5th edition:

https://www.skatteverket.se/download/18.361dc8c15312eff6fd10583/1461569090868/vat-on-foreign-trade-skv560b-utgava05.pdf

Reply:
Please use the luhn algorithm inside Odoo:

from odoo.tools.misc import mod10r

mod10r(orgnumber[:-1]) == orgnumber

Reply:
MMmmh. So it does state that VAT suffix is always 01. So, is Wikipedia wrong then ?

Quote from Wikipedia:
> 12 digits, of which the last two are most often 01 e.g. SE999999999901. (For sole proprietors who have several businesses the numbers can be 02, 03 and so on, since sole proprietors only have their personnummer as the organisationsnummer. The first 10 digits are the same as the Swedish organisationsnummer.[12]

And that point has been discussed on the talk tab of the wiki:
> When I follow the source, and read in Swedish, I see, that this is only for "enkla bolag". My experience tells me that "01" is the most common last digits among Swedish companies, but other numbers do exist for other kinds of companies, especially those who has an office on several locations. -- Lavallen (talk) 15:08, 24 April 2013 (UTC)

Reply:
@hpr-odoo  It would be indeed be a good idea to reuse existing code. Though I think `odoo.tools.misc.mod10r` makes hypotheses on the length of the input which don't match our requirements. Maybe you can make it length agnostic using your own implementation so that you can use it in every situation.

Here's my test:
```python
def verify(number):
    return mod10r(number[:-1]) == number

print(verify('8112289874'))  # False
print(verify('6709199530'))  # False
print(verify('8512289870'))  # False
print(verify('79927398713'))  # True
```

If need be, I found another implementation online which works in every case: https://github.com/mmcloughlin/luhn/blob/master/luhn.py .

Reply:
I haven't found references to 02 except on the wikipedia page. However 'Enkla Bolag' isn't really a company and can't have a registred orgnumber and it doesn't need to do any bookkeeping. Since each indiviual in 'Enkla Bolag' is responsible for the economic transactions. If they did bookkeeping I bet they use K1 rules and since this is a K2 account table it doesn't fit into the scope.

And by the way, the 'Enkla Bolag' can't have it's own VAT-nummer. The VAT-number is in that case based on the individual and not the 'Enkla Bolag'. Personal number and organisational numbers have the same format. In that case you are doing business with a private person and not a company since 'Enkla Bolag' can't be a juridical person.

K1, Individual traders, non-profit associations, registered faith communities and Handelsbolag (HB)
K2, Small Companies, Aktiebolag (AB), Economic associations (EF)
K3, Group of Companies

Reply:
Here is the code from /odoo/tools/misc.py:

def mod10r(number):
    """
    Input number : account or invoice number
    Output return: the same number completed with the recursive mod10
    key
    """
    codec=[0,9,4,6,8,2,7,1,3,5]
    report = 0
    result=""
    for digit in number:
        result += digit
        if digit.isdigit():
            report = codec[ (int(digit) + report) % 10 ]
    return result + str((10 - report) % 10)

Reply:
Your right.. It calculates it wrong.. 

```
def validate_luhn(acc_number):
    n = len(acc_number)
    digits = [int(d) for d in re.sub(r'\D', '', acc_number)][-n:]
    even_digitsum = sum(x if x < 5 else x - 9 for x in digits[::2])
    return 0 == sum(digits, even_digitsum) % 10
```
Works better.

Reply:
Ok, thanks for the clarification. `01` it is then.

Reply:
done

Reply:
I think luhn is working as expect. i have tested all numbers which you had tried to verify. luhn could validate it successfully.

>>> from stdnum import luhn
>>> luhn.validate('1234567897')
'1234567897'
>>> luhn.validate('8112289874')
'8112289874'
>>> luhn.validate('6709199530')
'6709199530'
>>> luhn.validate('8512289870')
'8512289870'
>>> luhn.validate('79927398713')
'79927398713'

@AdrienHorgnies , please correct me if i am wrong.